nio/client/async_client.py

```diff
@@ -3160,3 +3174,131 @@ async def set_pushrule_actions(
+                                         m.room.avatar, m.room.topic, m.room.guest_access,
+                                         m.room.history_visibility, m.room.join_rules, m.room.power_levels
+
+            room_upgrade_message (str): Message inside the tombstone-event
+
+            room_power_level_overwrite (dict): A ``m.room.power_levels content`` dict
+                to override the default.
+                The dict will be applied on top of the generated
+                ``m.room.power_levels`` event before it is sent to the room.
+        """
+
+        # Get state events for the old room
+        old_room_state_events = await self.room_get_state(old_room_id)
+        if isinstance(old_room_state_events, RoomGetStateError):
+            return RoomUpgradeError("Failed to get room events")
+
+        # Get initial_state and power_level
+        old_room_power_levels = None
+        new_room_initial_state = list()
+        for event in old_room_state_events.events:
+            if event['type'] in copy_events \
+               and not event['type'] == 'm.room.power_levels':
+                new_room_initial_state.append(event)
+            if event['type'] == 'm.room.power_levels':
+                old_room_power_levels = event['content']
+
+        # Get last known event from the old room
+        old_room_event = await self.room_messages(
+                                                start="",
+                                                room_id=old_room_id,
+                                                limit=1
+        )
+        if isinstance(old_room_event, RoomMessagesError):
+            return RoomUpgradeError("Failed to get last known event")
+
+        old_room_last_event = old_room_event.chunk[0]
+
+        # Overwrite power level if a new power level was passed
+        if room_power_level_overwrite is not None:
+            old_room_power_levels = room_power_level_overwrite
+
+        # Create new room
+        new_room = await self.room_create(
+            room_version=new_room_version,
+            power_level_override=old_room_power_levels,
+            initial_state=new_room_initial_state,
+            predecessor={
+                "event_id": old_room_last_event.event_id,
+                "room_id": old_room_id
+            })
+
+        if isinstance(new_room, RoomCreateError):
+            return RoomUpgradeError("Room creation failed")
+
+        # Send tombstone event to the old room
+        old_room_tombstone = await self.room_put_state(old_room_id, "m.room.tombstone",
+                                                       {"body": room_upgrade_message,
+                                                        "replacement_room": new_room.room_id
+                                                       })
+        if isinstance(old_room_tombstone, RoomPutStateError):
+            return RoomUpgradeError("Failed to put m.room.tombstone")
```

Comment:
Yes, is there an easy option to do that? Or does that involve more changes to the code base, if so I'd favor to move permission-checking into another PR as this one is getting quite huge already. As long as that's okay for everyone

Reply:
Well, it's relatively straightforward:
 - you have to figure out who you are using the `whoami` endpoint
 - you have to check what your PL is
 - you have to check what the PL required for sending the state event is
 - if that isn't set, you need to fall back for the state event default
 
 It probably makes sense to put that in a separate method as well, but I'd prefer to have that in here anyway, because as it currently is, we do lots of stuff that might already cause changes in the rooms, and then we end up not tombstoning the room, because we don't have any permissions to do that.